osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/SecretManagementService.java

```diff
@@ -1,181 +1,359 @@
-                    SecretType.E_METER_ENCRYPTION_KEY_UNICAST);
+            return this.aesEncryptionService.decrypt(encryptedKey);
+        } catch (final Exception e) {
+            throw new FunctionalException(FunctionalExceptionType.ENCRYPTION_EXCEPTION, ComponentType.PROTOCOL_DLMS, e);
+        }
+    }
 
-            byte[] decryptedKey = decryptSoapSecret(deviceIdentification, optionalTypedSecret.orElseThrow(
-                    () -> new IllegalStateException("Secret not found:" + deviceIdentification)));
+    @Override
+    public byte[] getKey(String deviceIdentification, SecurityKeyType keyType) {
+        LOGGER.info("Retrieving {} for device {}", keyType.name(), deviceIdentification);
+        return this.getKeys(deviceIdentification, Arrays.asList(keyType)).get(keyType);
+    }
 
-            log.trace("DlmsGlobalUnicastEncryptionKey for device " + deviceIdentification + " is " + Hex.encodeHexString(decryptedKey));
+    @Override
+    public Map<SecurityKeyType, byte[]> getKeys(String deviceIdentification, List<SecurityKeyType> keyTypes) {
+        GetSecretsRequest request = this.createGetSecretsRequest(deviceIdentification, keyTypes);
+        GetSecretsResponse response = this.secretManagementClient.getSecretsRequest(request);
+        this.validateGetResponse(keyTypes, response);
+        return this.convertSoapSecretsToSecretMapByType(response.getTypedSecrets().getTypedSecret());
+    }
 
-            return decryptedKey;
+    @Override
+    public byte[] getNewKey(String deviceIdentification, SecurityKeyType keyType) {
+        LOGGER.info("Retrieving new {} for device {}", keyType.name(), deviceIdentification);
+        return this.getNewKeys(deviceIdentification, Arrays.asList(keyType)).get(keyType);
+    }
 
-        } catch (Exception e) {
-            LOGGER.error("Error while retrieving encryption key", e);
+    @Override
+    public Map<SecurityKeyType, byte[]> getNewKeys(String deviceIdentification, List<SecurityKeyType> keyTypes) {
+        GetNewSecretsRequest request = this.createGetNewSecretsRequest(deviceIdentification, keyTypes);
+        GetNewSecretsResponse response = this.secretManagementClient.getNewSecretsRequest(request);
+        this.validateGetNewResponse(keyTypes, response);
+        return this.convertSoapSecretsToSecretMapByType(response.getTypedSecrets().getTypedSecret());
+    }
+
+    private void validateGetResponse(List<SecurityKeyType> keyTypes, GetSecretsResponse response) {
+        if (!OsgpResultType.OK.equals(response.getResult()) || response.getTypedSecrets() == null
+                || response.getTypedSecrets().getTypedSecret() == null) {
+            throw new IllegalStateException(
+                    "Invalid/incomplete soap response: resulttype=" + response.getResult().value());
+        }
+        if (keyTypes.size() != response.getTypedSecrets().getTypedSecret().size()) {
+            throw new IllegalStateException("Unexpected number of secrets in response");
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Define a constant instead of duplicating this literal "Unexpected number of secrets in response" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1192)